numpy/core/src/multiarray/mapping.c

```diff
@@ -228,129 +215,11 @@ unpack_scalar(PyObject *index, PyObject **result, npy_intp NPY_UNUSED(result_n))
 NPY_NO_EXPORT npy_intp
 unpack_indices(PyObject *index, PyObject **result, npy_intp result_n)
 {
-    npy_intp n, i;
-    npy_bool commit_to_unpack;
-
-    /* Fast route for passing a tuple */
-    if (PyTuple_CheckExact(index)) {
-        return unpack_tuple((PyTupleObject *)index, result, result_n);
-    }
-
-    /* Obvious single-entry cases */
-    if (0  /* to aid macros below */
-            || PyLong_CheckExact(index)
-            || index == Py_None
-            || PySlice_Check(index)
-            || PyArray_Check(index)
-            || !PySequence_Check(index)
-            || PyUnicode_Check(index)) {
-
-        return unpack_scalar(index, result, result_n);
-    }
-
-    /*
-     * Passing a tuple subclass - coerce to the base type. This incurs an
```

Comment:
I think this changes whether `__getitem__` is called on tuple subclasses; something that I don't think was part of the original deprecation

Reply:
Would you prefer I revert the first part of this function to be like https://github.com/numpy/numpy/blob/main/numpy/core/src/multiarray/mapping.c#L231-L263

and then follow that with
```
return unpack_scalar(index, result, result_n);
```
?

(That's roughly how I had it before.)

Reply:
Hmmm, I was worried that it might, but then noticed that
```
class myt(tuple):
    def __getitem__(self, i):
        return i
m = myt([1, 2, 3])
tuple(m)
```
prints, `(1, 2, 3)`, so `__getitem__` is not called for `tuple(obj)`.

However, I missed that if you also implement `__iter__` things *will* change and `PySequence_Tuple` seems to worry about it.  So maybe we should just keep calling it...  (A test would be nice if we care enough, even if it has a comment: "Probably doesn't matter")

I might consider reverse the logic in the code to `if (!PyTuple_Check()) {...}` in that case.  And we maybe might as well in-line the scalar case as it only occurs once now.

Reply:
Yes, I think custom `__iter__` was what I had in mind when I wrote this a while ago. I can't remember whether I actually cared about that behavior, or simply was trying to preserve the behavior we had before. At any rate, I think we should preserve it in this PR that finalizes the deprecation, and can rip it out later if we decide it matters.

Reply:
I reverted this to the more conservative version of the change. Please take a look!

Reply:
Thanks, but I really want those "Obvious single-entry cases" checks gone, they serve no purpose if we are limited to tuples and tuple subclasses.

Reply:
That is why I said we probably might just lead the whole thing with a `if (!PyTuple_Check(...))` for the "scalar" path.  Then comes the exact tuple, then that odd subclass support.
(But exact order here is probably depends on personal preferences.)

Reply:
OK with me, I was trying to order it from most to least likely, but I doubt it seems likely to not even make a measurable difference for `obj_arr1d[0]` so it doesn't matter.